app/models/user.rb

```diff
@@ -0,0 +1,17 @@
+class User < ApplicationRecord
+  before_save :downcase_email
+  validates :name, presence: true,
+    length: {maximum: Settings.user_validates.name_max}
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true,
+    length: {maximum: Settings.user_validates.email_max},
+    format: {with: VALID_EMAIL_REGEX},
+    uniqueness: true
+  has_secure_password
```

Comment:
cái này là gì thế, tác dụng của nó là gì =))

Reply:
bcrypt là một thuật toán mã hóa tạo hash. has_secure_password là một method của rails sử dụng thuật toán bcrypt để tạo mã hash ạ. Sử dụng bcrypt trong app thì cần khai báo thêm gem có tên Bcrypt. Hash là một cách mã hóa một chiều giúp bảo mật thông tin, password là một thông tin nhạy cảm cần được bảo mật nên cần được tạo hash để an toàn ạ <(")